### PR TITLE
✨Refactor Solution Explorer

### LINKED ISSUE #498: 🐛 Doppelklick geht nur, wenn Studio noch nicht offen ist.
Ein Doppelklick importiert ein Diagram nur, wenn das Studio noch nicht offen ist.

Neueste Version.

### LINKED ISSUE #544: ✨ Add Refresh Button to Solution Explorer
#### Describe your issue

Currently there is no way to refresh the solution explorer. In VSCode there are multiple actions that can be performed:

grafik

#### How can others reproduce the issue?

- Open File SolutionExplorer
- Add new files in opened folder 
- New files dont show up in studio

#### Possible solution

Add a button to refresh the solution explorer

#### Issue checklist

Please check the boxes in this list after submitting your Issue:

- [x] I've checked if this issue already exists
- [x] I've included all the information that i think is relevant
- [x] I've added logs and/or screenshots (if applicable)
- [x] I've mentioned PRs and issues that relate to this one

### PR DESCRIPTION
**Pull request by Vyperus**
## What did you change?

This branch will:
- Update solution explorer dependencies.
- Add the double click feature.
- Refactor the old version of this feature.
- Close electron when all windows are closed.
- Fix a display bug in which the process in the navbar was not updated.
- Add a section for single diagrams in the solution explorer.
- Add a refresh button to the solution explorer to refresh the solutions if needed.
- Add a button to open single diagrams to the solution explorer.
- Add a button to open solutions to the solution explorer.
- Generally enhances the style of the solution explorer.

BPMN-Studio will now:
- Open a double-clicked file with the file-system-solution-explorer-service, display it and show it in the solution explorer under "Opened diagrams".
- Do the same as above when it's already open.

## Related Issues

Closes #498. 
Closes #544.

## How can others test the changes?

- Checkout this branch.
- `npm run electron-build-macos`
- Wait.
- Open the folder with: `open .`.
- Navigate to `dist/mac/`.
- Start BPMN-Studio App.
- Double click a .bpmn-file of your choice.

## PR-Checklist

Please check the boxes in this list after submitting your PR:

- [x] You can merge this PR **right now** (if not, please prefix the title with "WIP: ")
- [x] I've tested **all** changes included in this PR.
- [x] I've also reviewed this PR myself before submitting (e.g. for scrambled letters, typos, etc.)
- [x] I've merged the `develop` branch into my branch before finishing this PR.
- [x] I've **not added any other changes** than the ones described above.
- [x] I've mentioned all **PRs, which relate to this one**

### CODE REVIEW

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 1]
```diff
@@ -40,9 +41,46 @@ export class ProcessSolutionPanel {
...
+
+      // Check if there was a double click before BPMN-Studio was loaded.
+      const fileInfo: IFileInfo = ipcRenderer.sendSync('get_opened_file');
+
+      if (fileInfo.path) {
+        const solutionPath: string = fileInfo.path.substr(0, fileInfo.path.lastIndexOf('/'));
+
+        await this._solutionExplorerServiceFileSystem.openSolution(solutionPath, this._identity);
+        const solution: ISolution = await this._solutionExplorerServiceFileSystem.loadSolution();
+        const diagramName: string = fileInfo.path.substring(fileInfo.path.lastIndexOf('/') + 1, fileInfo.path.length - this._bpmnSuffixLength);
```
Comment (Paulomart) [inline src/modules/process-solution-panel/process-solution-panel.ts]: You could make use of `path.basename` https://nodejs.org/api/path.html#path_path_basename_path_ext

File: src/modules/diagram-detail/diagram-detail.ts [hunk 2]
```diff
@@ -5,13 +5,14 @@ import {Redirect, Router} from 'aurelia-router';
 import {IDiagram} from '@process-engine/solutionexplorer.contracts';
 import {ISolutionExplorerService} from '@process-engine/solutionexplorer.service.contracts';
 
+import { IIdentity } from '../../../node_modules/@process-engine/bpmn-studio_client';
```
Comment (robpal94) [inline src/modules/diagram-detail/diagram-detail.ts]: Please remove the whitespaces 🙂

File: src/modules/process-solution-panel/process-solution-panel.scss [hunk 3]
```diff
@@ -8,11 +8,48 @@
...
+}
+
+.control-buttons {
+  float: right;
+}
+
+.control-buttons button {
+  border: none;
+  background: transparent;
+  padding: 0 2px;
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.scss]: Please fix the order:

.control-buttons button {
  padding: 0 2px;
  background: transparent;
  border: none;
}

https://github.com/5minds/css-guidelines

File: src/modules/process-solution-panel/process-solution-panel.scss [hunk 4]
```diff
@@ -8,11 +8,48 @@
...
+}
+
+.control-buttons button:hover {
+  color: #5e5e5e;
+}
+
+.single-diagram-input {
+  visibility: hidden;
+  height: 0;
+}
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.scss]: .single-diagram-input {
  height: 0;
  visibility: hidden;
}

File: src/modules/process-solution-panel/process-solution-panel.scss [hunk 5]
```diff
@@ -8,11 +8,48 @@
...
+
+.single-diagram-close {
+  float: right;
+  position: relative;
+  top: 3px;
+  color: darkgray;
+}
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.scss]: .single-diagram-close {
  position: relative;
  top: 3px;
  float: right;
  color: darkgray;
}

File: src/modules/process-solution-panel/process-solution-panel.scss [hunk 6]
```diff
@@ -46,6 +81,9 @@
 
 .solution-close {
   padding-left: 10px;
+  color: darkgray;
+  position: relative;
+  top: 1px;
 }
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.scss]: .solution-close {
  position: relative;
  top: 1px;
  padding-left: 10px;
  color: darkgray;
}

File: src/modules/processdef-list/processdef-list.ts [hunk 7]
```diff
@@ -62,29 +61,8 @@ export class ProcessDefList {
...
     this._eventAggregator.publish(environment.events.refreshProcessDefs);
-
-    this._fileReader.onload = async(fileInformations: any): Promise<void> => {
-      const xml: string = fileInformations.target.result;
-      const filename: string = this.selectedFiles[0].name;
-
-      this.fileInput.value = '';
-
-      this._importXmlFromFile(filename, xml);
-    };
```
Comment (SteffenKn) [inline src/modules/processdef-list/processdef-list.ts]: This is needed for the "Import Process"-Button.

File: src/modules/processdef-list/processdef-list.ts [hunk 8]
```diff
@@ -62,29 +61,8 @@ export class ProcessDefList {
...
-    if (fileHasContent) {
-      // This Regex cuts out the filename from the filepath.
-      const filename: string = /[^\\/:*?"<>|\r\n]+$/.exec(fileInfo.path)[0];
-      const xml: string = fileInfo.content;
-
-      fileInfo.content = undefined;
-
-      this._importXmlFromFile(filename, xml);
-    }
-
```
Comment (SteffenKn) [inline src/modules/processdef-list/processdef-list.ts]: This is needed for the "Import Process"-Button.

File: src/modules/processdef-list/processdef-list.ts [hunk 9]
```diff
@@ -36,7 +36,6 @@ export class ProcessDefList {
   private _getProcessesIntervalId: number;
   private _newDiagramXml: string;
 
-  @bindable()
```
Comment (SteffenKn) [inline src/modules/processdef-list/processdef-list.ts]: This is necessary for the "Import Process"-Button.

File: src/modules/process-solution-panel/process-solution-panel.scss [hunk 10]
```diff
@@ -8,11 +8,48 @@
...
+}
+
 .process-explorer__heading {
   color: #a5a5a5;
   font-weight: 400;
-  padding-left: 10px;
-  padding-top: 10px;
+  padding: 8px;
   margin-bottom: 10px;
 }
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.scss]: .process-explorer__heading {
  padding: 8px;
  margin-bottom: 10px;
  font-weight: 400;
  color: #a5a5a5;
}

File: src/modules/process-solution-panel/process-solution-panel.scss [hunk 11]
```diff
@@ -23,7 +60,6 @@
 
 .process-explorer__structure-item a {
   color: #5c5c5c;
-  display: block;
   padding-left: 10px;
 }
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.scss]: .process-explorer__structure-item a {
  padding-left: 10px;
  color: #5c5c5c;
}

File: src/modules/process-solution-panel/process-solution-panel.ts [hunk 12]
```diff
@@ -98,13 +120,31 @@ export class ProcessSolutionPanel {
...
       return solution.uri === solutionToClose.uri;
     });
     this.openedFileSystemSolutions.splice(index, 1);
   }
 
+  public closeSingleDiagram(diagramToClose: IDiagram): void {
+    const index: number = this.openedSingleDiagrams.findIndex((diagram: IDiagram) => {
+      return diagram.uri === diagramToClose.uri;
+    });
+    this.openedSingleDiagrams.splice(index, 1);
```
Comment (SteffenKn) [inline src/modules/process-solution-panel/process-solution-panel.ts]:
```	
	const index: number = this.openedSingleDiagrams.findIndex((diagram: IDiagram) => {
      return diagram.uri === diagramToClose.uri;
    });

    this.openedSingleDiagrams.splice(index, 1);
```

Please add the whitespace to increase readability.

**Pull request merged**